src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -803,6 +817,13 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
                 return false;
             }
 
+            if (typeParameter.HasUnmanagedTypeConstraint && (typeArgument.IsManagedType || !typeArgument.IsNonNullableValueType()))
```

Comment:
>typeArgument.IsManagedType || !typeArgument.IsNonNullableValueType() [](start = 61, length = 68)

It might be better to check these two requirements independently and to report separate errors for them. For example for the IsNonNullableValueType we would report the same error as in the ```if```below and for the typeArgument.IsManagedType we would say about members. Alternatively we, I think it would be better to say "must be non-nullable value type" instead of saying "cannot be a reference type", unless that is inaccurate.

Reply:
I prefer the second option. I think it is great to have a separate full error message for `unmanaged` constraint for developers to mentally differentiate.

---
In reply to: [172414637](https://github.com/dotnet/roslyn/pull/24817#discussion_r172414637) [](ancestors = 172414637)